[web-animations] support CSSNumberish? values for Animation.startTime and Animation.currentTime setters

graouts opened the pull request:
#### 4dc7e1ab89bb9ca1ec5863f324583758f933bba0
<pre>
[web-animations] support CSSNumberish? values for Animation.startTime and Animation.currentTime setters
<a href="https://bugs.webkit.org/show_bug.cgi?id=246876">https://bugs.webkit.org/show_bug.cgi?id=246876</a>

Reviewed by Chris Dumez.

Change the currentTime and startTime attributes of Animation to be CSSNumberish? instead of double?
per the latest Web Animations 2 draft (<a href="https://drafts.csswg.org/web-animations).">https://drafts.csswg.org/web-animations).</a>

Note that the WPT tests were incorrect in expecting a NotSupportedError exception instead of TypeError
when validating the numberish value as specified by <a href="https://drafts.csswg.org/web-animations/#validating-a-css-numberish-time.">https://drafts.csswg.org/web-animations/#validating-a-css-numberish-time.</a>

* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-current-time-of-an-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-current-time-of-an-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation.html:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::setBindingsStartTime):
* Source/WebCore/animation/CSSAnimation.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::validateCSSNumberishValue const):
(WebCore::WebAnimation::setBindingsStartTime):
(WebCore::WebAnimation::setBindingsCurrentTime):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/animation/WebAnimation.idl:

Canonical link: <a href="https://commits.webkit.org/255870@main">https://commits.webkit.org/255870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47d2a78fdffe4649b858ed78c348adec73444936

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103484 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163817 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97844 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3060 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31279 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86171 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99519 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2181 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80275 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29201 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84107 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72154 "Found 2 new API test failures: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37675 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17638 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35539 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18900 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39418 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41471 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1911 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41353 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38130 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -301,20 +306,55 @@ void WebAnimation::effectTargetDidChange(const std::optional<const Styleable>& p
 }
 
+#if ENABLE(CSS_TYPED_OM)
+ExceptionOr<std::optional<Seconds>> WebAnimation::validateCSSNumberishValue(std::optional<CSSNumberish> optionalCSSNumberish) const
```
Should probably use `const std::optional<CSSNumberish>&` to avoid refcounting churn when CSSNumberish contains a RefPtr.

graouts replied:
Filed https://github.com/WebKit/WebKit/pull/5644 to fix a similar problem that had already snuck into `CSSScale`.

cdumez commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -301,20 +306,55 @@ void WebAnimation::effectTargetDidChange(const std::optional<const Styleable>& p
 }
 
+#if ENABLE(CSS_TYPED_OM)
+ExceptionOr<void> WebAnimation::setBindingsStartTime(std::optional<CSSNumberish> startTime)
```
`const std::optional<CSSNumberish>&` or `std::optional<CSSNumberish>&&`.

cdumez commented on Source/WebCore/animation/WebAnimation.idl:
```diff
@@ -50,8 +54,13 @@ typedef unsigned long FramesPerSecond;
     attribute DOMString id;
     [ImplementedAs=bindingsEffect] attribute AnimationEffect? effect;
     [EnabledConditionallyReadWriteBySetting=WebAnimationsMutableTimelinesEnabled] attribute AnimationTimeline? timeline;
+#if defined(ENABLE_CSS_TYPED_OM) && ENABLE_CSS_TYPED_OM
```
We should probably kill this build time flag sooner rather than later. It is starting to make the code look pretty ugly and we have a runtime flag already.

graouts replied:
Yes, filed https://github.com/WebKit/WebKit/pull/5649 to address that.

cdumez reviewed (approved):
LGTM

graouts committed:
[web-animations] support CSSNumberish? values for Animation.startTime and Animation.currentTime setters
https://bugs.webkit.org/show_bug.cgi?id=246876

Reviewed by Chris Dumez.

Change the currentTime and startTime attributes of Animation to be CSSNumberish? instead of double?
per the latest Web Animations 2 draft (https://drafts.csswg.org/web-animations).

Note that the WPT tests were incorrect in expecting a NotSupportedError exception instead of TypeError
when validating the numberish value as specified by https://drafts.csswg.org/web-animations/#validating-a-css-numberish-time.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-current-time-of-an-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-current-time-of-an-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation.html:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::setBindingsStartTime):
* Source/WebCore/animation/CSSAnimation.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::validateCSSNumberishValue const):
(WebCore::WebAnimation::setBindingsStartTime):
(WebCore::WebAnimation::setBindingsCurrentTime):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/animation/WebAnimation.idl:

Canonical link: https://commits.webkit.org/255870@main
